corehq/apps/api/odata/serializers.py

```diff
@@ -29,39 +30,43 @@ def to_json(self, data, options=None):
         self.offset = data.get('meta', {}).get('offset', 0)
 
         # Convert bundled objects to JSON
-        data['objects'] = [
-            bundle.obj for bundle in data['objects']
-        ]
-
-        domain = data.pop('domain', None)
-        config_id = data.pop('config_id', None)
-        api_path = data.pop('api_path', None)
-        table_id = data.pop('table_id', None)
-
-        assert all([domain, config_id, api_path]), [domain, config_id, api_path]
-
-        context_urlname = self.metadata_url
-        context_url_args = [domain, config_id]
-        if table_id > 0:
-            context_urlname = self.table_metadata_url
-            context_url_args.append(table_id)
-
-        data['@odata.context'] = '{}#{}'.format(
-            absolute_reverse(context_urlname, args=context_url_args),
-            'feed'
-        )
-
-        next_link = self.get_next_url(data.pop('meta'), api_path)
-        if next_link:
-            data['@odata.nextLink'] = next_link
-
-        config = self.get_config(config_id)
-        data['value'] = self.serialize_documents_using_config(
-            data.pop('objects'),
-            config,
-            table_id
-        )
-        return json.dumps(data, cls=DjangoJSONEncoder, sort_keys=True)
+        try:
+            data['objects'] = [
+                bundle.obj for bundle in data['objects']
+            ]
+
+            domain = data.pop('domain', None)
+            config_id = data.pop('config_id', None)
+            api_path = data.pop('api_path', None)
+            table_id = data.pop('table_id', None)
+
+            assert all([domain, config_id, api_path]), [domain, config_id, api_path]
+
+            context_urlname = self.metadata_url
+            context_url_args = [domain, config_id]
+            if table_id > 0:
+                context_urlname = self.table_metadata_url
+                context_url_args.append(table_id)
+
+            data['@odata.context'] = '{}#{}'.format(
+                absolute_reverse(context_urlname, args=context_url_args),
+                'feed'
+            )
+
+            next_link = self.get_next_url(data.pop('meta'), api_path)
+            if next_link:
+                data['@odata.nextLink'] = next_link
+
+            config = self.get_config(config_id)
+            data['value'] = self.serialize_documents_using_config(
+                data.pop('objects'),
+                config,
+                table_id
+            )
+            return json.dumps(data, cls=DjangoJSONEncoder, sort_keys=True)
+
+        except KeyError as e:
+            raise Http404(e)
```

Comment:
Instead of passing an error object, do you think it would be a good idea to pass a custom error message?

Reply:
Also, in order to raise a better custom error message, as @AmitPhulera, maybe this part of the OData spec helps: https://docs.oasis-open.org/odata/odata-json-format/v4.0/os/odata-json-format-v4.0-os.html#_Toc372793091

Can you return a response formatted like what's described in that doc and see how PowerBI deals with it?

Reply:
Yeah `data['objects']` doesn't exist and raised a 500.